Add data center documentation (#85)
GSA/automate.fedramp.gov: content/documentation/ssp/4-ssp-template-to-oscal-mapping.md

{{</callout>}}

> ---

Author: Resolved, see: 570af5c1a5ff1c6e651fc6a143a9c954a8474070

## Data Centers

Each system must define at least two data centers. There must be exactly one primary data center, and there must be at least one alternate data center. Additionally, the country specified in the data center's address must be within the United States.

#### OSCAL Representation
{{< highlight xml "linenos=table" >}}
<metadata>
    <!-- role -->
    <location uuid="uuid-of-primary-data-center">
       <title>Primary Data Center</title>
       <address>
          <addr-line>1234 Some Street</addr-line>
          <city>Haven</city>
          <state>ME</state>
          <postal-code>00000</postal-code>
          <country>us</country>
       </address>

>        <prop name='data-center' value='some-location-value' class='primary' ns="https://fedramp.gov/ns/oscal"/>

Author: Thank you for the explanation. Resolved, see: 570af5c1a5ff1c6e651fc6a143a9c954a8474070

    </location>
    <location uuid="uuid-of-alternate-data-center">
       <title>Secondary Data Center</title>
       <address>
          <addr-line>5678 Some Street</addr-line>
          <city>Haven</city>
          <state>ME</state>
          <postal-code>00000</postal-code>
          <country>us</country>
       </address>

>        <prop name='data-center' value='some-location-value' class='alternate' ns="https://fedramp.gov/ns/oscal"/>

Author: Resolved: 570af5c1a5ff1c6e651fc6a143a9c954a8474070

    </location>
    <!-- party -->
</metadata>
{{</ highlight >}}

#### XPath Queries
{{< highlight xml "linenos=table" >}}

>     Number of Data Centers:
>         count(/*/metadata/location[prop[@value eq 'data-center']]) > 1
>     Number of Primary Data Centers:
>         count(/*/metadata/location/prop[@value eq 'data-center'][@class eq 'primary']) = 1
>     Number of Alternate Data Centers:
>         count(/*/metadata/location/prop[@value eq 'data-center'][@class eq 'alternate']) > 0
>     Data Center Country:
>         /*/metadata/location[prop[@value eq 'data-center']]/address/country eq 'US'

Reviewer: Need to key off of the `prop` name

```suggestion
    Number of Data Centers:
        count(/*/metadata/location[prop[@name eq 'data-center']]) > 1
        count(/*/metadata/location[prop[@name eq 'data-center']]) > 1
    Number of Primary Data Centers:
        count(/*/metadata/location/prop[@name eq 'data-center'][@class eq 'primary']) = 1
    Number of Alternate Data Centers:
        count(/*/metadata/location/prop[@name eq 'data-center'][@class eq 'alternate']) > 0
    Data Center Country:
        /*/metadata/location[prop[@name  eq 'data-center']]/address/country = 'US'
```

Author: > Need to key off of the `prop` name

Thank you for catching that. Resolved, see: 570af5c1a5ff1c6e651fc6a143a9c954a8474070

        /*/metadata/party[@uuid=/*/metadata/party[@uuid=[/*/metadata/responsible-party[@role-id="information-system-security-officer"]/party-uuid]]/member-of-organization]/name
{{</ highlight >}}

---

## Data Centers

> Each system must define at least two data centers. There must be exactly one primary data center, and there must be at least one alternate data center. Additionally, the country specified in the data center's address must be within the United States.

Reviewer: The example shows "US" but if we don't explicitly callout that they must use [ISO 3166 Alpha-2](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2), people are likely to input other things that will not pass our validation checks (e.g., "United States", "USA", "usa", "us").

```suggestion
Each system must define at least two data centers. There must be exactly one primary data center, and there must be at least one alternate data center. Additionally, the country specified in the data center's address must be the United States, and must be in [ISO 3166 Alpha-2](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2) two-letter country code format (e.g., "US" in all upper case).
```

All other changes look great!!!

Reviewer: btw, sorry for not submitting this change with prior review...I had start do but between meetings must have submitted my review without posting this suggestion

> Each system must define at least two data centers. There must be exactly one primary data center, and there must be at least one alternate data center. Additionally, the country specified in the data center's address must be the United States, and must be in [ISO 3166 Alpha-2](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2) two-letter country code format (e.g., "US" in all upper case).

Reviewer 2: Sorry to come and be annoying because I am a few minutes behind, perhaps point to the official OSCAL model docs that explained this to me, not Wikipedia? I did that as a tl;dr casually among developers, and much of the data is exhaustive. Or maybe it is useful, I don't know.

```suggestion
Each system must define at least two data centers. There must be exactly one primary data center, and there must be at least one alternate data center. Additionally, the country specified in the data center's address must be the United States, and must be in [ISO 3166 Alpha-2](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2) two-letter country code format (e.g., "US" in all upper case).

```suggestion
Each system must define at least two data centers. There must be exactly one primary data center, and there must be at least one alternate data center. Additionally, the country specified in the data center's address must be the United States. It must be in [ISO 3166 Alpha-2 format](https://pages.nist.gov/OSCAL-Reference/models/v1.1.2/system-security-plan/xml-reference/#/system-security-plan/metadata/location/address/country) two-letter country code format (e.g., "US" in all upper case).
```

#### OSCAL Representation
{{< highlight xml "linenos=table" >}}
<metadata>
    <!-- role -->
    <location uuid="uuid-of-primary-data-center">
       <title>Primary Data Center</title>
       <address>
          <addr-line>1234 Some Street</addr-line>
          <city>Haven</city>
          <state>ME</state>
          <postal-code>00000</postal-code>

>           <country>us</country>

Reviewer 2: Blocking: [ISO 3166 Alpha-2](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2), which this constraint requires (and is explained in the upstream OSCAL documentation from NIST), ought to be capitalized, not lower case. If you run these examples through the constraints test harness it should fail.

```suggestion
          <country>US</country>
```

       </address>
       <prop name="data-center" value="some-location-value" class="primary"/>
    </location>
    <location uuid="uuid-of-alternate-data-center">
       <title>Secondary Data Center</title>
       <address>
          <addr-line>5678 Some Street</addr-line>
          <city>Haven</city>
          <state>ME</state>
          <postal-code>00000</postal-code>

>           <country>us</country>

Reviewer 2: Blocking: [ISO 3166 Alpha-2](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2), which this constraint requires (and is explained in the upstream OSCAL documentation from NIST), ought to be capitalized, not lower case. If you run these examples through the constraints test harness it should fail.

```suggestion
          <country>US</country>
```
